extension/api.js

```diff
@@ -0,0 +1,189 @@
+  getAPI(context) {
+    const { extension } = context;
+
+    return {
+      addonsSearchExperiment: {
+        // `getMatchPatterns()` returns a map where each key is an URL pattern
+        // to monitor and its corresponding value is an add-on ID (search
+        // engine).
+        //
+        // Note: We don't return a simple list of URL patterns because the
+        // background script might want to lookup the add-on ID for a given
+        // URL.
+        async getMatchPatterns() {
+          const patterns = {};
+
+          try {
+            await searchInitialized;
+            const visibleEngines = await Services.search.getVisibleEngines();
+
+            visibleEngines.forEach((engine) => {
+              let { _extensionID, _urls } = engine;
+
+              _urls.forEach(({ template }) => {
+                // If this is changed, double check the code in the background
+                // script because `webRequestCancelledHandler` splits patterns
+                // on `*` to retrieve URL prefixes.
+                const pattern = template.split("?")[0] + "*";
+                patterns[pattern] = _extensionID;
+              });
+            });
+          } catch (err) {
+            Cu.reportError(err);
+          }
+
+          return patterns;
+        },
+
+        // `getAddonVersion()` returns the add-on version if it exists.
+        async getAddonVersion(addonId) {
+          const addon = await AddonManager.getAddonByID(addonId);
+
+          return addon && addon.version;
+        },
+
+        // `getPublicSuffix()` returns the public suffix/Effective TLD Service
+        // of the given URL.
+        // See: https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIEffectiveTLDService
+        async getPublicSuffix(url) {
+          try {
+            return eTLD.getBaseDomain(Services.io.newURI(url));
+          } catch (err) {
+            Cu.reportError(err);
+            return null;
+          }
+        },
+
+        // `onSearchEngineModified` is an event that occurs when the list of
+        // search engines has changed, e.g., a new engine has been added or an
+        // engine has been removed.
+        //
+        // See: https://searchfox.org/mozilla-central/source/toolkit/components/search/SearchUtils.jsm#145-152
+        onSearchEngineModified: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onSearchEngineModified",
+          register: (fire) => {
+            const onSearchEngineModifiedObserver = (
+              aSubject,
+              aTopic,
+              aData
+            ) => {
+              if (
+                aTopic !== SEARCH_TOPIC_ENGINE_MODIFIED ||
+                // We are only interested in these modified types.
+                !["engine-added", "engine-removed"].includes(aData)
+              ) {
+                return;
+              }
+
+              searchInitialized.then(() => {
```

Comment:
I thought that your goal was to skip initial search engine modifications? The logic here wouldn't skip them, but delay the event dispatch.

Check for the `Services.search.isInitialized` if you want to check a boolean value in order to skip if needed.

Reply:
The goal was to make sure the search service is initialized before we can create and use an event in the background script. We don't necessarily need to skip any logic. I guess `searchInitialized` might not even be needed here because it is also used in `getMatchPatterns()`.